helix-rest/src/main/java/org/apache/helix/rest/metadatastore/ZkMetadataStoreDirectory.java

```diff
@@ -170,20 +170,21 @@ public void refreshRoutingData(String namespace) {
     // Check if namespace exists; otherwise, return as a NOP and log it
     if (!_routingZkAddressMap.containsKey(namespace)) {
       LOG.error("Failed to refresh internally-cached routing data! Namespace not found: " + namespace);
+      return;
     }
 
     try {
-      _realmToShardingKeysMap.put(namespace, _routingDataReaderMap.get(namespace).getRoutingData());
+      Map<String, List<String>> routingData = _routingDataReaderMap.get(namespace).getRoutingData();
+      _realmToShardingKeysMap.put(namespace, routingData);
+
+      if (_routingDataMap != null) {
+        MetadataStoreRoutingData newRoutingData = new TrieRoutingData(routingData);
+        _routingDataMap.put(namespace, newRoutingData);
+      }
     } catch (InvalidRoutingDataException e) {
-      LOG.error("Failed to get routing data for namespace: " + namespace + "!");
+      LOG.error("Routing data construction has failed for namespace: " + namespace + "!");
```

Comment:
Not necessarily. There are a few different places where you could get this exception, so your log could be misleading.

You should do `LOG.error("Failed to refresh cached routing data for namespace {}, namespace, e);` This propagates the correct exception message.

Reply:
Good point. I wasn't sure how to phrase this message because there are multiple reasons. I'll use your advised one.